core/src/main/java/dk/alexandra/fresco/framework/network/async/AsyncNetwork.java

```diff
@@ -0,0 +1,303 @@
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This network functions asynchronously in that sending happens within a thread, meaning that the
+ * method returns immediately after calling send. Receiving is blocking, but threads listen for
+ * incoming messages and passes them to the blocking queues where the main thread is potentially
+ * waiting.
+ */
+public class AsyncNetwork implements Network, Closeable {
+
+  private static final Logger logger = LoggerFactory.getLogger(AsyncNetwork.class);
+
+  ServerSocketChannel server;
+  Map<Integer, SocketChannel> clients;
+  Map<Integer, BlockingQueue<byte[]>> queues;
+  NetworkConfiguration conf;
+  ExecutorService receiverService;
+  Map<Integer, ExecutorService> senderServices;
+
+  /**
+   * Creates a network with the given configuration and a default timeout of 15 seconds. Calling the
+   * constructor will automatically trigger an attempt to connect to the other parties.
+   *
+   * @param conf The network configuration
+   */
+  public AsyncNetwork(NetworkConfiguration conf) {
+    this(conf, 15000);
+  }
+
+  /**
+   * Creates a network with the given configuration and a timeout of <code>timeout</code> in
+   * milliseconds. Calling the constructor will automatically trigger an attempt to connect to the
+   * other parties.
+   *
+   * @param conf The network configuration
+   * @param timeout
+   */
+  public AsyncNetwork(NetworkConfiguration conf, int timeout) {
+    this.conf = conf;
+    this.clients = new HashMap<>();
+    this.queues = new HashMap<>();
+    int fixedSize = Math.max(1, getNoOfParties() - 1);
+    this.receiverService = Executors.newFixedThreadPool(fixedSize);
+    this.senderServices = new HashMap<Integer, ExecutorService>();
+    for (int i = 1; i <= conf.noOfParties(); i++) {
+      if (this.conf.getMyId() != i) {
+        this.senderServices.put(i, Executors.newFixedThreadPool(1));
+      }
+    }
+    SocketAddress sock = new InetSocketAddress(conf.getParty(conf.getMyId()).getPort());
+    try {
+      this.server = ServerSocketChannel.open();
+      this.server.bind(sock);
+      logger.info("Bound at " + sock);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to bind to " + sock, e);
+    }
+    for (int i = 1; i <= conf.noOfParties(); i++) {
+      this.queues.put(i, new ArrayBlockingQueue<>(1000000));
```

Comment:
It's not so nice to have this as a literal constant. I would prefer having it as a final static field, say `DEFAULT_QUEUE_SIZE`. Maybe this value should even be configurable? You could also use a `LinkedBlockingQueue` (I think that is the name) to avoid having to cap the size.